recognition/include/pcl/recognition/impl/implicit_shape_model.hpp

```diff
@@ -192,8 +192,8 @@ pcl::features::ISMVoteList<PointT>::findStrongestPeaks (
         best_density = peak_densities[i];
         strongest_peak = peaks[i];
         best_peak_ind = i;
+        ++peak_counter;
```

Comment:
If the first peak in peak_densities is the highest value you would only get a peak_counter value of 1? But if its outside the if, you will get the a higher number , ie the actually number of peaks?

And how does it silence a warning about strongest_peak - I can't reason that :smile: ?

Reply:
ie. best_density is update on line 193, so its only first time its value is -1 ?

Reply:
It is only checked whether `peak_counter` is 0 or not (line 199), so whether it is 1 or a higher number does not matter.

The warning is that `strongest_peak` may be uninitialized. It is used in lines 203-205, and declared (but not initialized) in line 182. It is only initialized in line 193 if the test `peak_densities[i] > best_density` is true. Initially, `best_density` is -1 and we know that all `peak_densities` are non-negative, so this test will always be true when checked for the first time. However, the compiler/static analyzer does not know that all `peak_densities` are non-negative, so it does not know that this test will be true at least once. When `++peak_counter` is moved inside the if, `peak_counter` will only be greater than zero if the test `peak_densities[i] > best_density` passed at least once, and `strongest_peak` and the other variables are set to something valid.
So to summarize: the code did not have any problem, but the compiler/static analyzer does not have an information that we, the programmers have. Moving the increment makes it clear that `strongest_peak` has a "good" value if `peak_counter` is greater than zero.

Reply:
Thanks for the explanation!